rasterio/_vsiopener.pyx

```diff
@@ -62,6 +63,8 @@ cdef int install_pyopener_plugin(VSIFilesystemPluginCallbacksStruct *callbacks_s
         callbacks_struct.read = <VSIFilesystemPluginReadCallback>pyopener_read
         callbacks_struct.write = <VSIFilesystemPluginWriteCallback>pyopener_write
         callbacks_struct.close = <VSIFilesystemPluginCloseCallback>pyopener_close
+        # callbacks_struct.read_dir = <VSIFilesystemPluginReadDirCallback>pyopener_read_dir
```

Comment:
@rouault it looks like sidecar discovery doesn't strictly require read_dir. Is that true? Does it depend on the driver?

Reply:
> doesn't strictly require read_dir. Is that true?

yes, as I mentionned in my gdal-dev email, possibly in a confusing way :-), ReadDir() isn't strictly necessary, *for side car files*, because of the GDAL_DISABLE_READDIR_ON_OPEN=EMPTY_DIR hack/functionality, so all core or driver code that *can* use papszSiblingFiles got from readdir must also implement a fallback to using Stat() or Open() when papszSiblingFiles is empty.
That said there are a number of raster driver thats rely on ReadDir() for their normal operations: Sentinel2, ADRG, AIGRID, KMLSuperOverlay, MFF, HKV, SAGA, WMS (for the cache of tiles), Zarr, WCS (for a cache) . There are also vector drivers that need it: OpenFileGDB, Shape, CSV, etc. So I'd suggest it should be implemented

Reply:
Thanks @rouault ! Support for Zarr, WMS, WCS, &c is way out of scope for this, but Shapefile is within.